File: src/net/sourceforge/reb4j/IPAddressExampleTest.java
package net.sourceforge.reb4j;

import java.util.regex.Pattern;


import org.junit.Test;


import junit.framework.Assert;


public class IPAddressExampleTest
{
	private final Regex oneDigitOctet = 
		CharClass.POSIX_DIGIT;
	private final Regex twoDigitOctet = 
		CharClass.range('1', '9').then(CharClass.POSIX_DIGIT);
	private final Regex oneHundredsOctet =
		Regex.literal("1").then(CharClass.POSIX_DIGIT.repeat(2, 2));
	private final Regex lowTwoHundredsOctet =
		Regex.literal("2").then(CharClass.range('0', '4')).then(CharClass.POSIX_DIGIT);
	private final Regex highTwoHundredsOctet =
		Regex.literal("25").then(CharClass.range('0', '5'));
	private final Regex octet =
		Regex.or(
				oneDigitOctet, 
				twoDigitOctet, 
				oneHundredsOctet, 
				lowTwoHundredsOctet, 
				highTwoHundredsOctet
			);
	private final Regex dottedDecimalIPAddress =
		octet.then(Regex.literal(".")).repeat(3, 3).then(octet);
	
	
	
	@Test
	public void testOneDigitOctet()
	{
		final Pattern pattern = oneDigitOctet.toPattern();
		for (int i = 0; i < 10; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("10", pattern.matcher("10").matches());
	}
	
	
	@Test
	public void testTwoDigitOctet()
	{
		final Pattern pattern = twoDigitOctet.toPattern();
		for (int i = 10; i < 100; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("100", pattern.matcher("100").matches());
		Assert.assertFalse("9", pattern.matcher("9").matches());
	}
	
	
	
	public void testOneHundredsOctet()
	{
		final Pattern pattern = oneHundredsOctet.toPattern();
		for (int i = 100; i < 200; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("200", pattern.matcher("200").matches());
		Assert.assertFalse("99", pattern.matcher("99").matches());
	}
	
	
	@Test
	public void testLowTwoHundredsOctet()
	{
		final Pattern pattern = lowTwoHundredsOctet.toPattern();
		for (int i = 200; i < 250; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("250", pattern.matcher("250").matches());
		Assert.assertFalse("199", pattern.matcher("199").matches());
	}
	
	
	@Test
	public void testHighTwoHundredsOctet()
	{
		final Pattern pattern = highTwoHundredsOctet.toPattern();
		for (int i = 250; i < 256; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("256", pattern.matcher("256").matches());
		Assert.assertFalse("249", pattern.matcher("249").matches());
	}
	
	
	@Test
	public void testOctet()
	{
		final Pattern pattern = octet.toPattern();
		for (int i = 0; i < 256; i++)
		{
			String s = Integer.toString(i);
			Assert.assertTrue(s, pattern.matcher(s).matches());
		}
		Assert.assertFalse("256", pattern.matcher("256").matches());
		Assert.assertFalse("a", pattern.matcher("a").matches());
		Assert.assertFalse("1111", pattern.matcher("1111").matches());
	}
	
	
	
	@Test
	public void testDottedDecimalIPAddress()
	{
		//System.out.println(dottedDecimalIPAddress);
		final Pattern pattern =
			dottedDecimalIPAddress.toPattern();
		Assert.assertTrue("1.2.3.4", pattern.matcher("1.2.3.4").matches());
		Assert.assertTrue("123.239.35.254", pattern.matcher("123.239.35.254").matches());
		Assert.assertFalse("1.256.3.4", pattern.matcher("1.256.3.4").matches());
		Assert.assertFalse("1.2.a.4", pattern.matcher("1.2.a.4").matches());
	}
}

File: src/test/scala/net/sourceforge/reb4j/scala/test/IPAddressExampleTest.scala
package net.sourceforge.reb4j.scala.test
import org.junit.runner.RunWith
import org.scalatest.junit.JUnitRunner
import org.scalatest.FunSuite
import org.scalatest.Suite
import net.sourceforge.reb4j.scala.charclass.CharClass._
import org.scalatest.matchers.ShouldMatchers
import net.sourceforge.reb4j.scala._

@RunWith(classOf[JUnitRunner])
class IPAddressExampleTest extends Suite with ShouldMatchers with Implicits
{
	import Group.Capture
	val oneDigitOctet = Perl.Digit
	val twoDigitOctet = range('1', '9') ~~ Perl.Digit
	val oneHundredsOctet = '1' ~~ (Perl.Digit repeat 2)
	val lowTwoHundredsOctet = '2' ~~ range('0', '4') ~~ Perl.Digit
	val highTwoHundredsOctet = "25" ~~ range('0', '5')
	val octet = oneDigitOctet||twoDigitOctet||oneHundredsOctet||lowTwoHundredsOctet||highTwoHundredsOctet
	val dottedDecimalIPAddress = 
		Capture(octet) ~~ '.' ~~ 
		Capture(octet) ~~ '.' ~~ 
		Capture(octet) ~~ '.' ~~ 
		Capture(octet)
	
	def testOneDigitOctet()
	{
		val pattern = oneDigitOctet.toPattern
		for (i <- 0 to 9)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (true)}
		for (i <- 10 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
	}
	
	def testTwoDigitOctet()
	{
		val pattern = twoDigitOctet.toPattern
		for (i <- 0 to 9)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
		for (i <- 10 to 99)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (true)}
		for (i <- 100 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
	}
	
	def testOneHundredsOctet()
	{
		val pattern = oneHundredsOctet.toPattern
		for (i <- 0 to 99)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
		for (i <- 100 to 199)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (true)}
		for (i <- 200 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
	}
	
	def testLowTwoHundredsOctet()
	{
		val pattern = lowTwoHundredsOctet.toPattern
		for (i <- 0 to 199)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
		for (i <- 200 to 249)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (true)}
		for (i <- 250 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString).matches() should be (false)}
	}
	
	def testHighTwoHundredsOctet()
	{
		val pattern = highTwoHundredsOctet.toPattern
		for (i <- 0 to 249)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (false)}
		for (i <- 250 to 255)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (true)}
		for (i <- 256 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (false)}
	}
	
	def testOctet()
	{
		val pattern = octet.toPattern
		for (i <- 0 to 255)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (true)}
		for (i <- 256 to 1000)
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (false)}
		for (i <- 'a' to 'z')
			withClue(pattern + " matching " + i.toString + ":")
			{pattern.matcher(i.toString()).matches() should be (false)}
	}
	
	
	def testDottedDecimalIPAddress()
	{
		val pattern = dottedDecimalIPAddress.toPattern
		withClue(pattern + " matching 1.2.3.4:") 
			{pattern.matcher("1.2.3.4").matches() should be (true)}
		withClue(pattern + " matching 123.239.35.254:") 
			{pattern.matcher("123.239.35.254").matches() should be (true)}
		withClue(pattern + " matching 1.256.3.4:") 
			{pattern.matcher("1.256.3.4").matches() should be (false)}
		withClue(pattern + " matching 1.2.a.4:") 
			{pattern.matcher("1.2.a.4").matches() should be (false)}
	}
	
	def testCapture()
	{
		val regex = dottedDecimalIPAddress.toRegex()
		val regex(first, second, third, fourth) = "5.99.123.251"
		withClue(regex + " capturing the octets of 5.99.123.251:")
		{
			first should be ("5")
			second should be ("99")
			third should be ("123")
			fourth should be ("251")
		}
	}
	
}